src/txn/txn.c

```diff
@@ -597,6 +597,141 @@ __wt_txn_release(WT_SESSION_IMPL *session)
+     * let the search point to the last version of the key and start traversing backwards to find
+     * out the satisfying record according the given timestamp. Any satisfying history store record
+     * is moved into data store and removed from history store. If none of the history store records
+     * satisfy the given timestamp, the key is removed from data store.
+     */
+    WT_ERR(__wt_hs_cursor_position(session, hs_cursor, hs_btree_id, &op->u.op_row.key, WT_TS_MAX));
+    WT_ERR(hs_cursor->get_key(hs_cursor, &hs_btree_id, hs_key, &hs_start_ts, &hs_counter));
+
+    /* Stop before crossing over to the next btree */
+    if (hs_btree_id != S2BT(session)->id)
+        goto err;
+
+    /*
+     * Keys are sorted in an order, skip the ones before the desired key, and bail out if we have
+     * crossed over the desired key and not found the record we are looking for.
+     */
+    WT_ERR(__wt_compare(session, NULL, hs_key, &op->u.op_row.key, &cmp));
+    if (cmp != 0)
+        goto err;
+
+    /*
+     * As part of the history store search, we never get an exact match based on our search criteria
+     * as we always search for a maximum record for that key. Make sure that we set the comparison
+     * result as an exact match to remove this key as part of rollback to stable. In case if we
+     * don't mark the comparison result as same, later the __wt_row_modify function will not
+     * properly remove the update from history store.
+     */
+    hs_cbt->compare = 0;
+
+    /* Get current value and convert to full update if it is a modify. */
+    WT_ERR(hs_cursor->get_value(hs_cursor, &hs_stop_ts, &durable_ts, &type_full, hs_value));
+    WT_ASSERT(session, hs_stop_ts == WT_TS_MAX);
+
+    /*
+     * If we found a history value that satisfied the given timestamp, add it to the update list.
+     * Otherwise remove the key by adding a tombstone.
+     */
+    if (commit) {
+        WT_ERR(__wt_upd_alloc_tombstone(session, &hs_upd, NULL));
+        hs_upd->durable_ts = hs_upd->start_ts = txn->durable_timestamp;
+        hs_upd->txnid = txn->id;
+    } else {
+        WT_ERR(__wt_upd_alloc(session, hs_value, WT_UPDATE_STANDARD, &upd, NULL));
+
+        upd->txnid = WT_TXN_NONE;
+        upd->durable_ts = durable_ts;
+        upd->start_ts = hs_start_ts;
+        __wt_verbose(session, WT_VERB_TRANSACTION,
+          "update restored from history store (txnid: %" PRIu64 ", start_ts: %s, durable_ts: %s",
+          upd->txnid, __wt_timestamp_to_string(upd->start_ts, ts_string[0]),
+          __wt_timestamp_to_string(upd->durable_ts, ts_string[1]));
+
+        /*
+         * Set the flag to indicate that this update has been restored from history store for the
+         * rollback of a prepared transaction.
+         */
+        F_SET(upd, WT_UPDATE_RESTORED_FOR_ROLLBACK);
+
+        /* If there are existing updates, append them after the new update. */
+        if (cbt->ins != NULL)
+            upd->next = cbt->ins->upd;
```

Comment:
How can we have existing updates if we have an uncommitted prepared updates on disk?

Reply:
There exists a prepared update that is aborted. If we don't append it, it is a leak. I added an assert to make sure that `upd->next` is not NULL.

Reply:
I don't understand here. Why we can have an aborted prepare update on the chain if we already have a prepared update on the page. Should that be a conflict in the first place when we try to add the prepare update on the update chain? If they are the same update chain, should the update be freed in eviction?

Reply:
The prepared update is added back to the chain when the page is read back to memory. During the transaction commit/rollback, these updates will be modified to update the durable timestamp or aborted. So in rollback scenario, there exists an update that is aborted.

Reply:
OK, can you add a comment here and assert that that there is only one update and the update we see is aborted?

Reply:
updated.